packages/theme-data/src/utils/extendTheme.js

```diff
@@ -6,8 +6,19 @@
  * Add additional roles to a theme
  */
 export default function extendTheme(theme, extension) {
+  // NOTE: We do *not* use Object spread in `reduce`, since it copies the entire
+  // object in every iteration and becomes a performance bottleneck. Instead, we
+  // copy `theme` once at the beginning and then apply changes to that same
+  // object instance `acc` using property assignment and `Object.assign`.
   return Object.keys(extension).reduce(
-    (acc, key) => ({ ...acc, [key]: { ...theme[key], ...extension[key] } }),
-    theme
+    (acc, key) => {
+      if (!acc[key]) {
+        acc[key] = extension[key];
+      } else {
+        acc[key] = Object.assign(acc[key], extension[key]);
+      }
+      return acc;
+    },
+    { ...theme }
```

Comment:
This can still just be `theme`, no?

Reply:
Then I would be mutating the original `theme` object, I think?